src/client.ts

```diff
@@ -3,8 +3,16 @@ import * as assert from 'assert';
 import safe from '@smartlyio/safe-navigation';
 
 type HeaderProp<H, Next> = H extends void ? Next : { headers: H } & Next;
-type QueryProp<Q, Next> = Q extends void ? Next : { query: Q } & Next;
-type BodyProp<B> = B extends void ? object : { body: B };
+type QueryProp<Q, Next> = true extends HasOnlyOptionalTypes<Q>
+  ? { query?: Q } | ({ query?: Q } & Next) | Next
+  : { query: Q } | ({ query: Q } & Next);
+type BodyProp<B> = B extends void ? void : { body: B };
+
+type HasOnlyOptionalTypes<O> = {
+  [K in keyof O]?: O[K];
+} extends O
+  ? true
+  : false;
```

Comment:
Could probably simplify this to
```suggestion
type HasOnlyOptionalTypes<O> = Partial<O> extends O ? true : false;
```